src/lib/homestore_backend/hs_homeobject.cpp

```diff
@@ -87,16 +86,41 @@ HSHomeObject::~HSHomeObject() {
     iomanager.stop();
 }
 
-void HSHomeObject::on_shard_meta_blk_found(homestore::meta_blk* mblk, sisl::byte_view buf, size_t size) {
-    std::string shard_info_str;
-    shard_info_str.append(r_cast< const char* >(buf.bytes()), size);
+void HSHomeObject::on_shard_meta_blk_found(homestore::meta_blk* mblk, sisl::byte_view buf) {
+    homestore::superblk< shard_info_superblk > sb;
+    sb.load(buf, mblk);
 
-    auto shard = deserialize_shard(shard_info_str);
-    shard.metablk_cookie = mblk;
+    bool pg_is_recovered = false;
+    {
+        std::scoped_lock lock_guard(_pg_lock);
+        pg_is_recovered = _pg_map.find(sb->placement_group) != _pg_map.end();
+    }
 
-    // As shard info in the homestore metablk is always the latest state(OPEN or SEALED),
-    // we can always create a shard from this shard info and once shard is deleted, the associated metablk will be
-    // deleted too.
-    do_commit_new_shard(shard);
+    if (pg_is_recovered) {
+        auto hs_shard = std::make_shared< HS_Shard >(sb);
+        add_new_shard_to_map(hs_shard);
+        return;
+    }
+
+    // There is no guarantee that pg info will be recovered before shard recovery
+    std::scoped_lock lock_guard(recovery_mutex_);
+    pending_recovery_shards_[sb->placement_group].push_back(std::move(sb));
+}
+
+void HSHomeObject::on_shard_meta_blk_recover_completed(bool success) {
+    // Find all shard with opening state and excluede their binding chunks from the HeapChunkSelector;
+    RELEASE_ASSERT(pending_recovery_shards_.empty(), "some shards is still pending on recovery");
+    std::unordered_set< homestore::chunk_num_t > excluding_chunks;
+    std::scoped_lock lock_guard(_pg_lock);
+    for (auto& pair : _pg_map) {
+        for (auto& shard : pair.second->shards_) {
+            if (shard->info.state == ShardInfo::State::OPEN) {
```

Comment:
Does this mean if we havent fill the shards and they are open and they are not reused after recovery ?

Reply:
no, they will be reused after recovery because all shards will be recovered from HS metablks and opening shards are included too. here logic is to excluded chunks which has opening shards from the heap chunkselector, so they will not be re-used to create shard util current shards is sealed.